File: lang/python/cxlib/idioms.py
"""v0.8.0 Layer-2 Pythonic idioms — opt-in sugar over the Layer-1 surface.

Per `spec/bindings.md` §3.1, Layer 2 wraps Layer 1 with host-idiomatic
conveniences.  Every Layer-2 expression has a **single, documented**
Layer-1 desugaring; `explain(expr)` returns it as a string.

Importing from `cxlib.idioms` is OPTIONAL — Layer-1 is the byte-identical
cross-binding contract and lives at `cxlib.code.Doc`.  Layer-2 is
strictly additive sugar; it never replaces or modifies Layer-1
semantics.

Examples
--------

    from cxlib.idioms import Doc

    doc = Doc.parse(open("users.cx", "rb").read())

    # Subscript with CXPath — desugars to Layer-1 `select_all`.
    active = doc["//user[@active=true]"]
    # ≡ doc.select_all("//user[@active=true]")

    # Division-operator iterator — desugars to `select_all`.
    for email in doc / "//user/@email":
        print(email)
    # ≡ for email in doc.select_all("//user/@email"):

    # Subscript-assignment desugars to `modify` + Set.
    new_doc = doc.clone()
    new_doc["//user[@id=1]/@name"] = "Alice"
    # ≡ new_doc = new_doc.modify("//user[@id=1]/@name", '[set "Alice"]')

    # `explain(...)` returns the desugaring for any Layer-2 expression.
    explain(doc / "//user") == 'doc.select_all("//user")'

Wasm-blocked items deferred
---------------------------

The Layer-2 surface specified in `spec/bindings.md` §3.1 includes
generator-style iteration plus list comprehensions over Nodes.  The
comprehension path requires CXPath-from-AST inference at host-language
level and ships in a later phase; the iterator/subscript pair landed
here covers the common path.
"""
from __future__ import annotations

from typing import Any, Iterator, Optional

from . import code as _code
from .code import Doc as _BaseDoc
from .code import Node


__all__ = ["Doc", "Node", "explain"]


class Doc(_BaseDoc):
    """`cxlib.code.Doc` plus Pythonic dunder sugar.

    Every dunder method documents its Layer-1 desugaring inline so
    `cxlib.idioms.explain` can read it back at runtime.
    """

    __slots__ = ()

    # `doc["//user[...]"]` ≡ `doc.select_all("//user[...]")`.
    def __getitem__(self, cxpath: str) -> list[Node]:
        if not isinstance(cxpath, str):
            raise TypeError(
                f"Doc[...] expects a CXPath string, got {type(cxpath).__name__}"
            )
        return self.select_all(cxpath)

    # `doc["//path"] = "value"` ≡ `doc.modify("//path", '[set "value"]')`.
    # Mutates in place — Layer-2 sugar over the Layer-1 pure-functional
    # `modify` (we swap out the underlying Document atomically).
    def __setitem__(self, cxpath: str, value: Any) -> None:
        if not isinstance(cxpath, str):
            raise TypeError(
                f"Doc[...] = ... expects a CXPath string, got {type(cxpath).__name__}"
            )
        action = f'[set {_format_value(value)}]'
        new_doc = self.modify(cxpath, action)
        # Swap the underlying state — preserves identity semantics for
        # in-place subscript-assign.  Layer-1 callers who want pure-
        # functional update use `Doc.modify` directly.
        # Bypass __slots__ + immutable façade by writing through the
        # parent's slot names.
        object.__setattr__(self, "_bytes", new_doc.bytes())
        object.__setattr__(self, "_doc", new_doc.document)

    # `doc / "//user"` ≡ iterator over `doc.select_all("//user")`.
    # Returns a generator so the consumer can `for x in doc / "//y":`.
    def __truediv__(self, cxpath: str) -> Iterator[Node]:
        if not isinstance(cxpath, str):
            raise TypeError(
                f"Doc / ... expects a CXPath string, got {type(cxpath).__name__}"
            )
        return iter(self.select_all(cxpath))

    # `iter(doc)` ≡ iterator over the document's root-level children
    # (NOT a CXPath query — Python-idiomatic "iterate over the
    # document").  Desugaring: `doc.root().children()` when the root is
    # an element, else an empty iterator.
    def __iter__(self) -> Iterator[Node]:
        root = self.root()
        if root is None:
            return iter(())
        return iter(root.children())

    def __contains__(self, cxpath: str) -> bool:
        """``"//user" in doc`` ≡ ``len(doc.select_all("//user")) > 0``."""
        if not isinstance(cxpath, str):
            return False
        return bool(self.select_all(cxpath))

    # Layer-2 conveniences with no dunder hook.
    def clone(self) -> "Doc":
        """Return an independent Doc sharing no mutable state with the
        receiver.  ≡ ``Doc.parse(self.bytes())``."""
        return Doc.parse(self.bytes())


def _format_value(value: Any) -> str:
    """Render a Python value as CX literal text for ``[set V]`` actions.

    Strings are double-quoted (any embedded `"` is JSON-escaped); ints,
    floats, bools, and None map to their CX surface forms.  Other types
    raise ``TypeError`` — Layer-2 deliberately stays narrow so the
    desugaring is unambiguous.
    """
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, (int, float)):
        return str(value)
    if isinstance(value, str):
        # Minimal CX-string escape: backslash + double-quote.
        escaped = value.replace("\\", "\\\\").replace('"', '\\"')
        return f'"{escaped}"'
    raise TypeError(
        f"cxlib.idioms: unsupported value type {type(value).__name__} for '[set]' — "
        f"Layer-2 sugar supports None/bool/int/float/str only.  Drop down "
        f"to Layer-1 `Doc.modify(focus, action)` for richer shapes."
    )


# ── explain() — read back the Layer-1 desugaring ────────────────────────────
#
# Per spec/bindings.md §3.1: `cxlib.idioms.explain(expr)` returns the
# equivalent Layer-1 call for any Layer-2 expression.  We can't
# introspect arbitrary Python expressions, so the public signature
# accepts a (doc, op, *args) triple — used by fixtures + LSP hovers.

def explain(*args: Any) -> str:
    """Return the Layer-1 desugaring for a Layer-2 operation.

    Calling conventions:

        explain(doc, "getitem", cxpath)        → 'doc.select_all(cxpath)'
        explain(doc, "setitem", cxpath, value) → 'doc.modify(cxpath, "[set V]")'
        explain(doc, "truediv", cxpath)        → 'iter(doc.select_all(cxpath))'
        explain(doc, "iter")                   → 'doc.root().children()'
        explain(doc, "contains", cxpath)       → 'bool(doc.select_all(cxpath))'
    """
    if len(args) < 2:
        raise TypeError("explain(doc, op, *args) requires at least 2 positional args")
    _doc, op, *rest = args
    if op == "getitem":
        (cxpath,) = rest
        return f'doc.select_all({cxpath!r})'
    if op == "setitem":
        cxpath, value = rest
        return f'doc.modify({cxpath!r}, "[set {_format_value(value)}]")'
    if op == "truediv":
        (cxpath,) = rest
        return f'iter(doc.select_all({cxpath!r}))'
    if op == "iter":
        return 'doc.root().children()'
    if op == "contains":
        (cxpath,) = rest
        return f'bool(doc.select_all({cxpath!r}))'
    raise ValueError(f"explain: unknown Layer-2 op {op!r}")

File: lang/rust/cxlib/src/idioms.rs
//! v0.8.0 Layer-2 Rust idioms — opt-in sugar over the Layer-1 surface.
//!
//! Per [`spec/bindings.md` §3.3](../../../spec/bindings.md) the Layer-2
//! surface for Rust is `Iterator<Item = Node>` combinators that
//! desugar to Layer-1 `select_all` / `find_all` calls. Importing from
//! `cxlib::idioms` is OPTIONAL — Layer-1 lives at `cxlib::code::Doc`
//! and is the byte-identical cross-binding contract.
//!
//! # Examples
//!
//! ```ignore
//! use cxlib::idioms::DocExt;
//! use cxlib::code::Doc;
//!
//! let doc = Doc::from_str("[doc [user [name Alice]] [user [name Bob]]]")?;
//!
//! // Iterator combinator — desugars to Layer-1 `select_all`.
//! let names: Vec<_> = doc.select_iter("//user/name")?
//!     .filter_map(|n| match n.body() {
//!         serde_json::Value::String(s) => Some(s),
//!         _ => None,
//!     })
//!     .collect();
//! // ≡ doc.select_all("//user/name")?.into_iter()...
//!
//! // find_iter is the name-only variant — no CXPath parse.
//! let users: Vec<_> = doc.find_iter("user").collect();
//! // ≡ doc.find_all("user").into_iter()...
//! ```
//!
//! # Desugaring
//!
//! Per `spec/bindings.md §3.3`, every Layer-2 expression has a
//! documented Layer-1 equivalent:
//!
//! | Layer 2                              | Layer 1                       |
//! |--------------------------------------|--------------------------------|
//! | `doc.select_iter("//x")?`            | `doc.select_all("//x")?.into_iter()` |
//! | `doc.find_iter("user")`              | `doc.find_all("user").into_iter()`  |
//! | `doc.eval("[?...]")?`                | `Doc::eval` (Layer-1)              |
//! | `doc.diagram()?`                     | `Doc::diagram` (Layer-1)           |
//! | `doc.tree()?`                        | `Doc::tree` (Layer-1)              |
//!
//! `explain()` returns the desugaring string for any Layer-2
//! operation; used by fixtures and LSP hovers.

use crate::code::{Doc, Node};

/// Extension trait that adds iterator-combinator sugar to `Doc`. Bring
/// this trait into scope (`use cxlib::idioms::DocExt;`) to enable
/// `select_iter` / `find_iter` on any `Doc`.
pub trait DocExt {
    /// CXPath selection as an `Iterator<Item = Node>`. Desugars to
    /// `self.select_all(cxpath)?.into_iter()`.
    fn select_iter(&self, cxpath: &str) -> Result<std::vec::IntoIter<Node>, String>;

    /// Name-only selection as an `Iterator<Item = Node>`. Desugars to
    /// `self.find_all(name).into_iter()`.
    fn find_iter(&self, name: &str) -> std::vec::IntoIter<Node>;

    /// Mermaid diagram builder API — drop-in for `Doc::diagram()`
    /// with explicit format selection. Currently only `"mermaid"`
    /// is supported; other formats route through
    /// the CLI tier.
    fn diagram_with(&self, format: &str) -> Result<String, String>;
}

impl DocExt for Doc {
    fn select_iter(&self, cxpath: &str) -> Result<std::vec::IntoIter<Node>, String> {
        Ok(self.select_all(cxpath)?.into_iter())
    }

    fn find_iter(&self, name: &str) -> std::vec::IntoIter<Node> {
        self.find_all(name).into_iter()
    }

    fn diagram_with(&self, format: &str) -> Result<String, String> {
        // Layer-1 Doc::diagram is hard-coded to "mermaid"; route
        // through the free function cx_code_diagram for explicit
        // format pass-through (so Layer-2 callers can request future
        // SVG/PNG once those tiers land).
        crate::cx_code_diagram(
            std::str::from_utf8(self.source_bytes())
                .map_err(|e| format!("DocExt::diagram_with: source not valid UTF-8: {}", e))?,
            format,
        )
    }
}

/// Return the Layer-1 desugaring for a Layer-2 operation. Mirrors the
/// `cxlib.idioms.explain(...)` helper in the Python binding. Used by
/// fixtures and LSP hovers; not a runtime requirement.
///
/// ```ignore
/// use cxlib::idioms::explain;
/// assert_eq!(
///     explain("select_iter", "//user"),
///     r#"doc.select_all("//user")?.into_iter()"#
/// );
/// ```
pub fn explain(op: &str, arg: &str) -> String {
    match op {
        "select_iter"  => format!(r#"doc.select_all({:?})?.into_iter()"#, arg),
        "find_iter"    => format!(r#"doc.find_all({:?}).into_iter()"#, arg),
        "diagram_with" => format!(r#"cxlib::cx_code_diagram(doc.source_bytes(), {:?})"#, arg),
        _ => format!("explain: unknown Layer-2 op {:?}", op),
    }
}
